Classification metrics overhaul: input formatting standardization (1/n)

tadejsv opened the pull request:
This PR is a spin-off from #4835. It should be merged before any other spin offs, as it provides a base for all of them

## What does this PR do?

### General (fundamental) changes

I have created a new `_input_format_classification` function (in `metrics/classification/utils`). The job of this function is to a) validate, and b) transform the inputs into a common format. This common format is a binary label indicator array: either ``(N, C)``, or ``(N, C, X)`` (only for multi-dimensional multi-class inputs). 

I believe that having such a "central" function is crucial, as it gets rid of code duplication (which was present in PL metrics before), and enables metric developers to focus on developing the metrics themselves, and not on standardizing and validating inputs.

The validation performed on the inputs basically makes sure that they fall into one of the possible input type cases, that the values are consistent with both the type of the inputs and the additional parameters set (e.g. that there is no label higher than ``num_classes`` in target). The docstrings (and the new "Input types" section in the documentation) give all the details about how the standardization and validation are performed.

Here I'll list the parameters of this function (many of which are also present on some metrics), and why I decided to use them:

- `threshold`: The probability threshold for binarizing binary and multi-label inputs.

- `num_classes`: number of classes. Used to either decide the ``C`` dimension of inputs, or, if this is already implicitly given, to ensure consistency between inputs and number of classes the user specified when creating the metric (thus ignoring either having to chech this manually in `update` for each metric, or raising error when updating the state, which may not be very clear
to the user).

- `top_k`: for (multi-dimensional) multi-class, if predictions are given as probabilities, selects the top k highest probabilities per sample. It's a generalization of the usual procedure, with `k=1`. This will be used by the `Accuracy` metric in subsequent PRs.
- `is_multiclass`: used for transforming binary or multi-label input to 2-class multi-class and 2-class multi-dimensional multi-class, respectively. And vice versa.

    *Why?* This is similar to `multilabel` argument that was (is?) present on some metrics. I believe this is a better name for it, as it also deals with transforming to/from binary. But why is it needed? There are cases where it is not clear what the inputs are: for example, say that both preds and target are of the form [0,1,0,1]. This actually appears to be multi-class (could be the case that is simply happened in this batch that there were only 0s and 1s), so an explicit instruction is needed to tell the metrics that this is in fact binary. On the other hand, sometimes we would like to treat binary inputs as two class inputs - this is the case used in confusion matrix.

    I also experiemented with using `num_classes` to determine this. Besides this being a very confusing approach, requiring several paragraphs to explain clearly, it also does not resolve all ambiguities (is setting `num_classes=1` with 2 class probability predictions a request to treat the data as binary, or an inconsitency of inputs that should raise an error?). So I think `is_multiclass` is the best approach here.

### Documentation

Instead of metrics being organized into "Class Metrics" and "Functional Metrics", they are now organized by topics (Classification, Regression, ...), and within topics split into class and functional, if necessary. This allows to add special topic-related sections - in this case I have added a section on what type of inputs are used for classification metric - a section that metrics can link to, in order to not repeat the same thing 100 times, and to keep docstrings short and to the point.

A second half of the Input types section with examples from StatScores metric will be added in the metric's PR.

teddykoker commented:
Thanks for splitting off the PR! Reviewing now

teddykoker commented on docs/source/metrics.rst:
```diff
@@ -196,12 +196,71 @@ Metric API
 
+The functional metrics follow the simple paradigm input in, output out. This means, they don't provide any advanced mechanisms for syncing across DDP nodes or aggregation over batches. They simply compute the metric value based on the given inputs.
+
+Also the integration within other parts of PyTorch Lightning will never be as tight as with the class-based interface.
```
Might want to re-word this

tadejsv replied:
Sure, have any ideas how? I did not actually write this - this is taken from the current documentation (if it were up to me it would be taken out entirely)

Borda replied:
@teddykoker do you have any suggestions?

teddykoker replied:
Hmm my bad, I'm not opposed to removing either, I think this just leads to more confusion

teddykoker commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,385 @@
+    dimension is ambiguous (e.g. if targets are a ``(7, 3)`` tensor, while predictions are a
+    ``(7, 3, 3)`` tensor), it will be assumed that the ``C`` dimension is the second dimension.
+    If this is not the case,  you should move it from the last to second place using
+    ``torch.movedim(preds, -1, 1)``.
```
`torch.movedim` not supported in minimal version

tadejsv replied:
Thanks! I'll still leave movedim in documentation for users that are on the newest pytorch, but I'll also mention permutedim for users with older versions

teddykoker commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,385 @@
+    else:
+        mode = "multi-dim multi-class"
+        if preds.shape[:-1] == target.shape:
+            preds = torch.movedim(preds, -1, 1)
```
`torch.movedim` not supported in minimum PyTorch version, tests are failing because of this. Perhaps we can change this to a transpose or permutedim

tadejsv replied:
Oh thanks - permute should work in this case.

teddykoker commented on tests/metrics/classification/test_inputs.py:
```diff
@@ -0,0 +1,301 @@
+
+
+def mvdim(x):
+    return torch.movedim(x, -1, 1)
```
see other comment on `torch.movedim`

teddykoker reviewed (changes requested):
Overall looks like good changes, just a few small things to fix.

rohitgr7 commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,385 @@
+        raise ValueError("If you set is_multiclass=False and preds are integers, then preds should not exceed 1.")
+
+    # Check that shape/types fall into one of the cases
+    if len(preds.shape) == len(target.shape):
```
```suggestion
    if preds.ndim == target.ndim:
```
you can use ndim everywhere instead of len(argument.shape). Just a bit cleaner.

tadejsv replied:
Good catch, thanks!

rohitgr7 commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,385 @@
+    # Check that shape/types fall into one of the cases
+    if len(preds.shape) == len(target.shape):
+        if preds.shape != target.shape:
+            raise ValueError("if preds and target have the same number of dimensions, they should have the same shape")
```
```suggestion
            raise ValueError(
                "preds and targets should have the same shape",
                f" got preds shape = {preds.shape} and target shape = {target.shape}."
            )
```
Just add the value, if possible, wherever an error is raised

tadejsv replied:
Cool, will add this (but using lazy string formatting)

tadejsv replied:
Eh, or without lazy string formatting - this is not logging anyway

SkafteNicki commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+    target: torch.Tensor,
+    threshold: float,
+    num_classes: Optional[int] = None,
+    is_multiclass: bool = False,
```
Is this argument necessary if we settle on num_classes = 1 means binary and num_classes!=1 is multiclass?

tadejsv replied:
Sooo right now the "division of labor" is that `num_classes` is there to determine the shape for one-hot transformation (if needed), and consistency/error checking - if you specified `num_classes` in init, this would make sure your data indeed has that many classes, as I think it is more graceful to catch the error here than when you are updating the state tensors.

And `is_multiclass` is there to determine if the inputs need to be transformed to other types. If we make `num_classes` do all the work, you may loose some error checking power: for example if you have, for some reason, ML data but with binary integer preds (this is one of the inputs in tests), then to signal that it is multi-label you would, without `is_multiclass`, need to pass `num_classes=1`. This is both a bit strange (number of classes is actually `C`), and prevents you from checking that number of classes is whatever was set at init.

tadejsv replied:
I've tried it wit `num_classes` doing everything before, but I've decided to add this extra parameter for the reasons above (and maybe I forgot something as well). Also, I think it's easier to explain this to the users, rather than adding some convoluted discussion on how `num_classes` can be used.

SkafteNicki replied:
I have to think a bit more about this, but here is a counter-argument:
Sklearn does not have this kind of argument, and therefore I believe it should absolutely be possible to differentiate between the different cases without. As a user, it feels strange that if I specify `num_classes=5` for example I also have to specify that `is_multiclass=True`. It seems redundant.

tadejsv replied:
There's no situation where you have to specify `num_classes=5` and `is_multiclass=True`

tadejsv replied:
`is_multiclass=True` is only unsed to convert binary and multi-label data into 2-class (multi-dimensional) multi-class

tadejsv replied:
Sklearn does not have this because, for example in recall, it has this extra average "binary", which users have to specify if they want to treat their data as binary. So they also can't differentiate between the classes automatically, but instead offload this to different parameters for each metric. I think my way is much better/clearer, as this is not about the the "averaging method", but about the nature of your data.

tadejsv replied:
Not to mention that sklearn does not deal with multi-dim multi-class data at all

SkafteNicki commented on docs/source/metrics.rst:
```diff
@@ -196,12 +196,71 @@ Metric API
+    "Multi-class with probabilities", "(N, C)", "``float``", "(N,)", "``int``"
+    "Multi-label", "(N, ...)", "``float``", "(N, ...)", "``binary``\*"
+    "Multi-dimensional multi-class", "(N, ...)", "``int``", "(N, ...)", "``int``"
+    "Multi-dimensional multi-class with probabilities", "(N, C, ...) or (N, ..., C)", "``float``", "(N, ...)", "``int``"
```
IMO we should only allow the first option: `(N, C, ...)`.
If this is specified in docs (like here, really like the table :] ) then it will up to the user to reshape

Borda replied:
I may agree here that all options would be quite confusing when code is going to be shared among users...

tadejsv replied:
Alright, if there seems to be a consensus on this I'll change it

Borda replied:
Let's ping also @justusschock and @teddykoker and @ananyahjha93

justusschock replied:
yes, just (N,C,...) is what I'd expect as well, since this is the standard tensor format in pytorch here. If user's are using anything else on purpose they should take care of proper formatting by themselves.

tadejsv replied:
Cool, whill change then

SkafteNicki reviewed (commented):
overall very good :)

SkafteNicki commented on docs/source/metrics.rst:
```diff
@@ -196,12 +196,71 @@ Metric API
+    All dimensions of size 1 (except ``N``) are "squeezed out" at the beginning, so 
+    that, for example, a tensor of shape ``(N, 1)`` is treated as ``(N, )``.
+
+When predictions or targets are integers, it is assumed that class labels start at , i.e. 
```
```suggestion
When predictions or targets are integers, it is assumed that class labels start at 0, i.e. 
```

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+        if preds.shape != target.shape and target.max() >= extra_dim_size:
+            raise ValueError("The highest label in targets should be smaller than the size of C dimension")
+    else:
+        if case == "binary":
+            if num_classes > 2:
+                raise ValueError("Your data is binary, but num_classes is larger than 2.")
+            elif num_classes == 2 and not is_multiclass:
+                raise ValueError(
+                    "Your data is binary and num_classes=2, but is_multiclass is not True."
+                    "Set it to True if you want to transform binary data to multi-class format."
+                )
+            elif num_classes == 1 and is_multiclass:
+                raise ValueError(
+                    "You have binary data and have set is_multiclass=True, but num_classes is 1."
+                    "Either leave is_multiclass unset or set it to 2 to transform binary data to multi-class format."
+                )
+        elif "multi-class" in case:
+            if num_classes == 1 and is_multiclass is not False:
+                raise ValueError(
+                    "You have set num_classes=1, but predictions are integers."
+                    "If you want to convert (multi-dimensional) multi-class data with 2 classes"
+                    "to binary/multi-label, set is_multiclass=False."
+                )
+            elif num_classes > 1:
+                if is_multiclass is False:
+                    if implied_classes != num_classes:
+                        raise ValueError(
+                            "You have set is_multiclass=False, but the implied number of classes "
+                            "(from shape of inputs) does not match num_classes. If you are trying to"
+                            "transform multi-dim multi-class data with 2 classes to multi-label, num_classes"
+                            "should be either None or the product of the size of extra dimensions (...)."
+                            "See Input Types in Metrics documentation."
+                        )
+                if num_classes <= target.max():
+                    raise ValueError("The highest label in targets should be smaller than num_classes")
+                if num_classes <= preds.max():
+                    raise ValueError("The highest label in preds should be smaller than num_classes")
+                if preds.shape != target.shape and num_classes != extra_dim_size:
+                    raise ValueError("The size of C dimension of preds does not match num_classes")
```
just thinking to wrap this into another protected function for:
- shorten the func as it is quite long and all the if/else are uneasy to follow
- give the functional block a name so easier to understand the whole context

tadejsv replied:
Ok, I can create some sub-functions here for each case

tadejsv replied:
Doneio :)

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+
+    preds_float = preds.is_floating_point()
+    if not preds_float and preds.min() < 0:
+        raise ValueError("if preds are integers, they have to be non-negative")
```
```suggestion
        raise ValueError("if `preds` are integers, they have to be non-negative.")
```
lets wrap all vars in `var_name` so it is easier to understand what is a general term and func var, in particular for `target`

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+    if preds.ndim == target.ndim:
+        if preds.shape != target.shape:
+            raise ValueError(
+                "preds and targets should have the same shape",
+                f" got preds shape = {preds.shape} and target shape = {target.shape}.",
```
```suggestion
                "The `preds` and `targets` should have the same shape",
                f" got `preds` shape = {preds.shape} and `target` shape = {target.shape}.",
```

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+            if preds.shape[2:] != target.shape[1:]:
+                raise ValueError(
+                    "if preds if preds have one dimension more than target, the shape of preds should be"
+                    "either of shape (N, C, ...) or (N, ..., C), and of targets of shape (N, ...)"
```
```suggestion
                    " either of shape (N, C, ...) or (N, ..., C), and of targets of shape (N, ...)"
```
if multiline, the next line starts with space otherwise there is no space between the last word on the previous line and fist word on new line...

tadejsv replied:
Done

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+
+    # Multi-dim multi-class (N, ...) with integers
+    elif preds.shape == target.shape and not preds_float:
+        mode = "multi-dim multi-class"
```
just an idea, shall we use , as a separator?
```suggestion
        mode = "multi-dim,multi-class"
```

tadejsv replied:
I like it better the way it is - corresponds more to the fully written out word (multi-dimensional multi-class)

rohitgr7 replied:
is it `multi-class (N, C, ...)`??

tadejsv replied:
In this case it is `(N, ...)`

rohitgr7 replied:
ok, it's for labels. :+1:

rohitgr7 replied:
I agree with Borda, should be separable by a separator.

Borda replied:
nothing bad, just a suggestion :]

Borda commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -36,8 +37,8 @@ def _flatten(x):
 def to_onehot(
-        tensor: torch.Tensor,
-        num_classes: int,
+    tensor: torch.Tensor,
```
```suggestion
    data: torch.Tensor,
```
can we name it better then tensor so it is not confusing with PT data type, well suggested `data` is also not a good one :]

tadejsv replied:
Changed it to `labels_tensor`

Borda commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -122,26 +139,24 @@ def _input_format_classification_one_hot(
+        raise ValueError("preds and target must have same number of dimensions, or one additional dimension for preds")
 
-    if len(preds.shape) == len(target.shape) + 1:
+    if preds.ndim == target.ndim + 1:
```
:+1:

Borda commented on tests/metrics/classification/test_inputs.py:
```diff
@@ -0,0 +1,306 @@
+T = torch.Tensor
+
+
+def idn(x):
```
if they are meant to be just helper function in this file/tests consider making then protected
```suggestion
def _idn(x):
```

Borda reviewed (commented):
overall I like it very much, just a few notes

tadejsv commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,390 @@
+        preds: binary tensor of shape (N, C) or (N, C, X)
+        target: binary tensor of shape (N, C) or (N, C, X)
+    """
+    preds, target = preds.clone().detach(), target.clone().detach()
```
But for classification metrics at least (which this function is for), would there even be any way to back-propagate? With the select-top-k and thresholding that is being done.

rohitgr7 replied:
https://discuss.pytorch.org/t/backward-through-topk-operation-on-variable/9197/2

tadejsv replied:
I see. Is there any possibility that would cause some graphs not to be destroyed between batches if this is not detached? Or does the `Metric` class handle state updates by detaching?

SkafteNicki replied:
It is described at the bottom of this section https://pytorch-lightning.readthedocs.io/en/stable/metrics.html#implementing-a-metric in the docs. Basically, right now you can backprop the current batch if you use forward:
```
val = metric(pred, target) # this value can be backpropagated
```
but not the accumulated metric value
```
val = metric.compute() # this value cannot be backpropagated
```

tadejsv replied:
Sorry, my question was more about the following:
Say you update the state (within a metric), so that
```
self.state += non_detached_tensor
```
Then, if I am not wrong, `self.state` also acquires `requires_grad=True` and is included in the computation graph. So is this state than detached, and if so, when does this happen?

SkafteNicki replied:
As the no grad happens outside `update`, everything inside should already have `requires_grad=False` so it should not be a problem I think
https://github.com/PyTorchLightning/pytorch-lightning/blob/f2fa3c82567ade0aa5bea3aa298542fc5040e4b7/pytorch_lightning/metrics/metric.py#L155-L156

pep8speaks commented:
Hello @tadejsv! Thanks for updating this PR. 

* In the file [`pytorch_lightning/metrics/classification/utils.py`](https://github.com/PyTorchLightning/pytorch-lightning/blob/a04a71ea195c601d59b22c7295c6d1389d7155fe/pytorch_lightning/metrics/classification/utils.py):

> [Line 98:9](https://github.com/PyTorchLightning/pytorch-lightning/blob/a04a71ea195c601d59b22c7295c6d1389d7155fe/pytorch_lightning/metrics/classification/utils.py#L98): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

justusschock commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,441 @@
+    preds, target = preds.clone().detach(), target.clone().detach()
+
+    # Remove excess dimensions
+    if preds.shape[0] == 1:
```
why do you first squeeze and then unsqueezed again? Isn't this the same as not doing anything at all?

tadejsv replied:
No. This is in case if data passed in is, for example, (1, 5, 1) - batch size 1, num_classes 5 and some useless extra dim (I remove this extra dims, as otherwise there would be waaaaay too many edge cases). In this case squeezing returns (5,), and unsqueezing gives (1,5)

justusschock replied:
But what if you have (10, 5, 1)? You still have the extra dim, but won't squeeze it?

tadejsv replied:
In this case it goes to the `else` statement below, which simply does the squeeze without any unsqueezing :)

tadejsv replied:
Unsqueezing is only in case the first dimension (batch size) equals 1

justusschock commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -56,25 +57,47 @@ def to_onehot(
+        tensor([[0, 1, 1],
+                [1, 1, 0]], dtype=torch.int32)
+    """
+    zeros = torch.zeros_like(prob_tensor, device=prob_tensor.device)
```
I think `torch.zeros_like(prob_tensor)` already uses `prob_tensor.device` and `prob_tensor.dtype` unless specified differently.
```suggestion
    zeros = torch.zeros_like(prob_tensor)
```

tadejsv replied:
Didn't know, thanks!

Borda commented on docs/source/metrics.rst:
```diff
@@ -196,12 +196,71 @@ Metric API
+When predictions or targets are integers, it is assumed that class labels start at 0, i.e. 
+the possible class labels are 0, 1, 2, 3, etc. Below are some examples of different input types
+
+.. code-block:: python
```
```suggestion
.. testcode::
```

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+    """
+    This checks that the shape and type of inputs are consistent with
+    each other and fall into one of the allowed input types (see the
+    documentation of docstring of _input_format_classification). It does
```
```suggestion
    documentation of docstring of ``_input_format_classification``). It does
```

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+        elif preds.ndim == 1 and not preds_float:
+            case = "multi-class"
+        elif preds.ndim > 1 and preds_float:
+            case = "multi-label"
```
```suggestion
            case = "multi-dim"
```
or how the "multi-label" differs from "multi-class"

tadejsv replied:
This terminology is take from [sklearn](https://scikit-learn.org/stable/modules/multiclass.html), let me copy the relevant parts:
- Multiclass classification: classification task with more than two classes. Each sample can only be labelled as one class.
- Multilabel classification: classification task labelling each sample with x labels from n_classes possible classes, where x can be 0 to n_classes inclusive. This can be thought of as predicting properties of a sample that are not mutually exclusive. Formally, a binary output is assigned to each class, for every sample. 

The same thing in [wikipedia](https://en.wikipedia.org/wiki/Multi-label_classification) [articles](https://en.wikipedia.org/wiki/Multiclass_classification)

Borda replied:
, can we include it in the docs, this explanation, eventually with link to wiki :]

Borda replied:
ok, I see it now in the docs... sorry :]

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+        mode = "binary"
+        preds = (preds >= threshold).int()
+
+        if is_multiclass:
+            target = to_onehot(target, 2)
+            preds = to_onehot(preds, 2)
+        else:
+            preds = preds.unsqueeze(-1)
+            target = target.unsqueeze(-1)
```
just thinking if we would here get just a lambda function which would be later, maybe at the func end applied to both like
```
_transfrom = lambda t: to_onehot(t, 2) if is_multiclass else lambda t: t.unsqueeze(-1)
```
and later, at the end
```
target = _transform(target)
preds = _transform(preds)
```
which would also guarantee that we apply the very same transformation to target and prediction and avoid typo

tadejsv replied:
There are cases where transformation for preds and target is not the same, and linter wouldn't like assigning lambdas :)

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+        mode = "multi-label"
+        preds = (preds >= threshold).int()
+
+        if is_multiclass:
+            preds = to_onehot(preds, 2).reshape(preds.shape[0], 2, -1)
+            target = to_onehot(target, 2).reshape(target.shape[0], 2, -1)
+        else:
+            preds = preds.reshape(preds.shape[0], -1)
+            target = target.reshape(target.shape[0], -1)
```
also, it could be, not sure if it is not too much engineering, we may define look-up-table where the key is the `mode` (plus extra level-key for `is_multiclass`) and value is the lambda transformation as above...

tadejsv replied:
I guess it could be done - I'll think about how to do it

tadejsv replied:
Well - on the second thought, I think this would obfuscate the code too much. Apart from `case` and `is_multiclass` it would also need to take into account number of classes (implicit or specified in `num_classes`, and for multi-class case it would need to take into account if `preds` were labels of probabilities (ok, this could be appended to `case`, but still).

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+
+    # Multi-dim multi-class (N, C, ...) and (N, ..., C)
+    else:
+        mode = "multi-dim multi-class"
```
just curious, does the order matter?

tadejsv replied:
What do you mean by order?

Borda replied:
"multi-dim multi-class" vs "multi-class multi-dim" but I remember that later you just check if the key in the string...

Borda commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -56,25 +57,47 @@ def to_onehot(
-    tensor_onehot = torch.zeros(shape[0], num_classes, *shape[1:],
-                                dtype=dtype, device=device)
-    index = tensor.long().unsqueeze(1).expand_as(tensor_onehot)
+    dtype, device, shape = label_tensor.dtype, label_tensor.device, label_tensor.shape
+    tensor_onehot = torch.zeros(shape[0], num_classes, *shape[1:], dtype=dtype, device=device)
```
```suggestion
    tensor_onehot = torch.zeros(
        label_tensor.shape[0],
        num_classes,
        *label_tensor.shape[1:],
        dtype=label_tensor.dtype,
        device=label_tensor.device,
    )
```

Borda reviewed (commented):
just minor comments :]

tadejsv commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,424 @@
+    )
+
+    preds_float = preds.is_floating_point()
+
```
Good point

tadejsv commented:
@Borda I made some changes to cut down the number of lines in formatting function down to absolute minimum - let me know if you prefer the new or the old version.

Borda commented:
> @Borda I made some changes to cut down the number of lines in formatting function down to absolute minimum - let me know if you prefer the new or the old version.

sorry if I confused you about nb lines, the main aspect is code readability... :]

tadejsv commented:
> > @Borda I made some changes to cut down the number of lines in formatting function down to absolute minimum - let me know if you prefer the new or the old version.
> 
> sorry if I confused you about nb lines, the main aspect is code readability... :]

Absolutely, that's why I ask whether you prefer the old or new version. Just looking for some feedback :)

rohitgr7 reviewed (commented):
@tarepan Looks good so far. Checks are strong :)

But `num_classes=1` seems a bit odd from the user's perspective to me.
Just some comments and suggestions for now. PR is big. Need to clarify the above point first.

tadejsv commented on docs/source/metrics.rst:
```diff
@@ -434,9 +440,57 @@ to_onehot [func]
 .. autofunction:: pytorch_lightning.metrics.functional.classification.to_onehot
     :noindex:
 
+******************
+Regression Metrics
+******************
```
Will do

tadejsv commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,382 @@
+        )
+    elif num_classes == 1 and is_multiclass:
+        raise ValueError(
+            "You have binary data and have set `is_multiclass=True`, but `num_classes` is 1."
+            " Either leave `is_multiclass` unset or set it to 2 to transform binary data to multi-class format."
```
By default `is_multiclass=None`. I see there's some misunderstanding about the use of this parameter - I'll write a longer explanation here (since it seems to be the only issue remaining) and improve the description in the docs.

rohitgr7 replied:
updated.
BTW you can go to `Files Changed` section and add multiple suggestions in a single commit

tadejsv replied:
Thanks, didn't know that :)

tadejsv commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,382 @@
+    and `is_multiclass` param for (multi-dimensional) multi-class data.
+    """
+
+    if num_classes == 1 and is_multiclass is not False:
```
So this is only for the case when the user wants to force multi-class data with 2 classes to binary. This can only happen if the user sets `is_multiclass=False` (that's why `is not False` here). Here `num_classes` is treated not as something that can change the "mode" of the data, but rather as a check on data - for example if your metric requires you to specify `num_classes` at init to allocate propely sized state tensors, this makes sure that inputs indeed conform to `num_classes` set then. But as I said, I'll write a bit more about `num_classes`

tadejsv commented on docs/source/metrics.rst:
```diff
@@ -196,12 +196,71 @@ Metric API
+Class vs Functional Metrics
+***************************
 
+The functional metrics follow the simple paradigm input in, output out. This means, they don't provide any advanced mechanisms for syncing across DDP nodes or aggregation over batches. They simply compute the metric value based on the given inputs.
+
+Also the integration within other parts of PyTorch Lightning will never be as tight as with the class-based interface.
```
I believe you are right. So the thing is that I didn't write this part - this is just taken from the current documentation (it's under Functional metrics right now). I think I rather leave rewording of this part to another PR :)

Also, for this suggestion: I think `input in, output out` is more descriptive than `input and output`, which was probably the intention of the author. Will add the comma tho :)

SkafteNicki replied:
@rohitgr7 it is true that metrics can be aggregated using `.log` but in some cases they will be aggregated wrongly.

rohitgr7 replied:
@SkafteNicki can you give an example?

SkafteNicki replied:
General problem:
Assuming that your number of samples `N` are not divisible by your batch size `B`, meaning that your last batch will be smaller than all the others. Since `.log` give uniform weight to all logged values when taking the average, the aggregated value will be biased towards the last batch.
A specific metric:
RMSE is one of the best examples, as it cannot be computed as a simple average over values calculated on individual batches (since `\sqrt(x+y)!=\sqrt(x)+\sqrt(y)`). To get this right, values inside the square root need to be aggregated first and the take the square root (which is what the metrics package does)

rohitgr7 replied:
oh nice!!

SkafteNicki commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -86,29 +114,23 @@ def _input_format_classification(
         preds = torch.argmax(preds, dim=1)
 
-    if len(preds.shape) == len(target.shape) and preds.dtype == torch.float:
+    if preds.ndim == target.ndim and preds.dtype == torch.float:
```
```suggestion
    if preds.ndim == target.ndim and preds.is_floating_point():
```

SkafteNicki commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -122,26 +144,24 @@ def _input_format_classification_one_hot(
         target = to_onehot(target, num_classes=num_classes)
 
-    elif len(preds.shape) == len(target.shape) and preds.dtype == torch.float:
+    elif preds.ndim == target.ndim and preds.dtype == torch.float:
```
```suggestion
    elif preds.ndim == target.ndim and preds.is_floating_point():
```

SkafteNicki commented on pytorch_lightning/metrics/utils.py:
```diff
@@ -122,26 +144,24 @@ def _input_format_classification_one_hot(
         preds = torch.argmax(preds, dim=1)
 
-    if len(preds.shape) == len(target.shape) and preds.dtype == torch.long and num_classes > 1 and not multilabel:
+    if preds.ndim == target.ndim and preds.dtype == torch.long and num_classes > 1 and not multilabel:
```
Allow for ints?
```suggestion
    if preds.ndim == target.ndim and preds.dtype in (torch.long, torch.int) and num_classes > 1 and not multilabel:
```

tadejsv replied:
Why not. By the time all my PRs are merged this function will become redundant anyway.

SkafteNicki commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,441 @@
+
+    if preds_float and (preds.min() < 0 or preds.max() > 1):
+        raise ValueError("The `preds` should be probabilities, but values were detected outside of [0,1] range.")
+
```
Should we check that `all(preds.sum(dim=1) == 1)` when user inputs probabilities?

tadejsv replied:
We could - though not for multi-label.

SkafteNicki commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,441 @@
+            default value (``None``) will be interepreted as 1 for these inputs.
+
+            Should be left unset (``None``) for all other types of inputs.
+        is_multiclass:
```
I am starting to understand the meaning of this parameter, and I am willing to include it for now to see how it plays out when actually integrated with some of our metrics. However, I really in doubt if `is_multiclass` is the correct name for it. If is see the argument name (without reading the documentation) I would suspect that I need to set it to `True` if I am running any form of multiclass classification

tadejsv replied:
Ok great :) What would you suggest as a name?

tadejsv replied:
Also, I prepared an essay about the use of `is_multiclass`, to better explain what it is meant for and why I think it is better than alternatives:

## How `is_multiclass` works

First, let me clear up some confusion about it: this is not something that needs to be set (i.e. changed from default `None`) in ~99% of cases. If the inputs fall into any of these cases (listed in the new table in documentation):
- binary (with probability predictions)
- (multi-dim) multi-class (with label or probability predictions)
- multi-label (with probability predictions)

the parameter *does not* need to be set. The new formatting function is "intelligent" enough to correctly deduce the type and apply proper formatting in these cases. I have updated the description of the parameter in the docstring to reflect this and avoid confusing users who don't need to use this parameter.

There are only 2 special cases where this parameter needs to be used. They concern cases where inputs appear as one case, but are actually another - and in a way that is ambiguous enough to make determining the actual case impossible without explicit user instruction, or where for some reason the user wants to treat inputs as a different case from what they appear. These cases are:

- **Use case 1**: for multi-class or multi-dim multi-class with 2 classes, if the user wants to treat inputs as binary or multi-label, respectively. To be explicit, here's what I mean

    ``` python
    # Appears as multi-class, is actually binary
    preds = [0,1,1,0,1]

    # Appears as multi-dim multi-class, is actually multi-label
    preds = [[0,1,0], [1,0,1], [0,0,1]]
    ```

    These 2 cases are currently (i.e. in master, before my PRs) included as test cases for metrics, so I assume that they are considered important enough to be supported.

    It's clear why the type can not be (correctly) automatically infered here: even if we assume that multi-class data with 2 classes does not exist in reality, it can be the case that this you have, for example, 3 classes in your data, but it just happens that only 2 were present in a batch.

- **Use case 2**: for binary or multi-label data, if the user wants to treat it as multi-class or multi-dim multi-class with 2 classes. This would be the case if the model gives prediction only for one class, but the user would like to compute the metrics for the implied 0, as well as the "default" 1 class. Here's what I mean (one example of binary data)

    ``` python
    preds = [0.1, 0.7, 0.2, 0.6]

    # without setting is_multiclass
    preds_transformed = [[0], [1], [0], [1]]

    # setting is_multiclass=True
    preds_transformed = [[1, 0], [0, 1], [1, 0], [1, 0]]
    ```

    An example of this would be image segmentation: if the prediction is binary (e.g. tissue is cancerous/not cancerous), usually only one class (foreground) is predicted. But the user might want to see the metrics for the "background" predictions as well. Sure, the model could be changed to output two values, the user could transform the outputs so that they contain 2 classes before feeding them into the metrics, etc. - but this all feels like extra "engineering code" that PL is supposed to do away with.

## Alternatives

Most importantly, here's why I chose to use ``is_multiclass`` and not the alternatives.

* **Alternative 1**: just use `num_classes`: This version would imply that the user wants to treat data a binary/multi-label if `num_classes=1`, and muli-class if `num_classes > 1`. There are a couple things I don't like about this alternative:

    - Some metrics (like recall, precision...), use the `num_classes` argument to determine the shape of the state variables. So setting `num_classes=1` to signal that data is actually multi-label (see the first use-case in the previous section), would cause the metrics to error out on the update step. To avoid these such metrics would then need to introduce another parameter (`num_classes_in_data`?). I think this alone disqualifies this alternative, but there are more reasons :)

    - This prevents `num_classes` from being used in data validation in the cases where it is used to override the aparent type. This then pushes errors down to the update step, where the error message (which will result, for example, from trying to add two tensors of incompatible sizes) will be much less clear to the user.

    - This prevents solving Use case 2 above: If multi-label data consists of 2 classes (`C` dimension equals 2), then setting `num_classes=2` can't specify whether the user wants to treat this as multi-dim multi-class data, or is just saying that this is multi-label data with 2 classes.

* **Alternative 2**: just use `num_classes` + abandon support for binary/multi-label predictions with labels. This will require users to manually cast these predictions to `float` before feeding them into the metrics. The prolems with this are

    - As mentioned before, binary/multi-label predictions with labels inputs are currently among the test cases, so I assume that it had been decided they must be supported.

    - Same as for alternative 1, Use case 2 can not be supported.


I think the only way we could possibly avoid using ``is_multiclass`` parameter is by requiring binary/multi-label inputs to be floats, and requiring users to transform single-class data to 2-class manually (for Use Case 2). But I think that overall this is not a good tradeoff.

tchaton replied:
Quick question about 
```
preds = [0.1, 0.7, 0.2, 0.6]

# without setting is_multiclass
preds_transformed = [[0], [1], [0], [1]]

# setting is_multiclass=True
preds_transformed = [[1, 0], [0, 1], [1, 0], [1, 0]]
```
Should it be 

```
preds_transformed = [[0, 1], [1, 0], [0, 1], [1, 0]] ?
```

tadejsv replied:
Is should actually be
```
preds_transformed = [[1, 0], [0, 1], [1, 0], [0, 1]]
```
If `prob>=0.5`, then this is class `1`, so `[0,1]`, if `prob < 0.5` it is class `0`, so `[1, 0]`.

SkafteNicki reviewed (commented):
Overall, I would say the PR is soon ready to be merged :)

tadejsv commented:
Is there anything else that needs to be done before this PR can be merged?
I think I've addressed all issues that arised so far, so we can bring the review process to an end - as I'm very anxious to finally get to the good stuff ;)

@Borda @SkafteNicki @teddykoker

SkafteNicki commented:
@tadejsv, thanks for the further description of the `is_multiclass` parameter. I am fine with merging this PR now, so please go ahead and solve the conflicts and I will approve it.

Borda commented:
@tadejsv mind resolve conflicts :] probably after #4549 
@SkafteNicki @justusschock @teddykoker @ananyahjha93 mind do the final review?

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,446 @@
+# Copyright The PyTorch Lightning team.
```
thinking about rename this module as "helpers"?
so we do not have too many utils in the metrics package :]

tadejsv replied:
Good idea yes

Borda commented on pytorch_lightning/metrics/classification/utils.py:
```diff
@@ -0,0 +1,446 @@
+        raise ValueError("If you set `is_multiclass=False` and `preds` are integers, then `preds` should not exceed 1.")
+
+
+def _check_shape_and_type_consistency(preds: torch.Tensor, target: torch.Tensor) -> Tuple[str, int]:
```
can we get tests for all these helper functions, just to be sure they work as expected... :]

tadejsv replied:
So these helper functions are tested as part of the bigger function - as they are only meant to be used with it. As you can see the coverage is 100% :)

Borda replied:
I see it now, overlooked it :]

Borda reviewed (approved):
LGTM, just would be nice to have tests also for all the helper functions raising some kind of exception...

tadejsv commented:
Alright, merge conflicts resolved, ready for final review. @SkafteNicki please double check that docs are ok (git diff not useful there).

SkafteNicki reviewed (approved):
LGTM, docs looks fine :]

awaelchli commented on docs/source/metrics.rst:
```diff
@@ -196,53 +196,76 @@ Metric API
-Class metrics
-*************
+***************************
+Class vs Functional Metrics
```
nice addtion! <3

awaelchli reviewed (approved):
High-level review, looks good, nice docs
